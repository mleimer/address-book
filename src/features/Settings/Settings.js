import React from 'react';
import {makeStyles} from '@material-ui/core/styles';
import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import FormControl from '@material-ui/core/FormControl';
import ListItemText from '@material-ui/core/ListItemText';
import Select from '@material-ui/core/Select';
import Checkbox from '@material-ui/core/Checkbox';
import Container from '@material-ui/core/Container';
import {useDispatch, useSelector} from 'react-redux';
import {NATIONALITIES} from '../../common/nationalities';
import {setNationalities} from '../../store/settings/settingsActions';
import Typography from '@material-ui/core/Typography';

const useStyles = makeStyles((theme) => ({
    container: {
        padding: '1rem'
    },
    formControl: {
        margin: theme.spacing(1),
        minWidth: 120,
        maxWidth: 300
    }
}));

const ITEM_HEIGHT = 48;
const ITEM_PADDING_TOP = 8;
const MenuProps = {
    PaperProps: {
        style: {
            maxHeight: ITEM_HEIGHT * 4.5 + ITEM_PADDING_TOP,
            width: 250
        }
    },
    getContentAnchorEl: null
};


function Settings() {

    const classes = useStyles();
    const selectedNationalities = useSelector(state => state.settings.nationalities);
    const dispatch = useDispatch();

    const handleChange = (event) => {
        dispatch(setNationalities(event.target.value));
    };

    return (
        <Container className={classes.container} data-testid="settings">
            <Typography component="h3">General filter settings</Typography>
            <FormControl className={classes.formControl}>
                <InputLabel id="nationalities-select-label">Nationalities</InputLabel>
                <Select
                    labelId="nationalities-select-label"
                    id="nationalities-select-checkbox"
                    multiple
                    value={selectedNationalities}
                    onChange={handleChange}
                    renderValue={(selected) => selected.join(', ')}
                    MenuProps={MenuProps}
                >
                    {NATIONALITIES.map((nationality) => (
                        <MenuItem key={nationality.key} value={nationality.key}>
                            <Checkbox checked={selectedNationalities.indexOf(nationality.key) > -1}/>
                            <ListItemText primary={nationality.value}/>
                        </MenuItem>
                    ))}
                </Select>
            </FormControl>
        </Container>
    );
}

Settings.propTypes = {};

Settings.defaultProps = {};

export default Settings;
