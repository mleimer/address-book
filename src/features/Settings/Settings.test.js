import React from 'react';
import {mockedStore, render} from '../../test/testUtils';
import Settings from './Settings';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';

describe('<Settings/>', () => {
    test('it should mount', () => {
        const {getByTestId} = render(
            <Settings/>
        );

        const settings = getByTestId('settings');

        expect(settings).toBeInTheDocument();
    });

    test('all nationalities should be pre-selected', () => {
        const {getByTestId} = render(
            <Settings/>
        );

        const nationalitiesSelect = getByTestId('nationalities-select');

        expect(nationalitiesSelect).toBeInTheDocument();
        expect(nationalitiesSelect).toHaveTextContent('CH, ES, FR, GB');
    });

    test('it should remain stable (snapshot test)', () => {
        const component = renderer.create(
            <Provider store={mockedStore}>
                <Settings/>
            </Provider>
        ).toJSON();

        expect(component).toMatchSnapshot();
    });

});
