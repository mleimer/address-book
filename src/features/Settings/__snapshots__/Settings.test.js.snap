// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Settings/> it should remain stable (snapshot test) 1`] = `
<div
  className="MuiContainer-root makeStyles-container-5 MuiContainer-maxWidthLg"
  data-testid="settings"
>
  <h3
    className="MuiTypography-root MuiTypography-body1"
  >
    General filter settings
  </h3>
  <div
    className="MuiFormControl-root makeStyles-formControl-6"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
      data-shrink={true}
      id="nationalities-select-label"
    >
      Nationalities
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      data-testid="nationalities-select"
      onClick={[Function]}
    >
      <div
        aria-haspopup="listbox"
        aria-labelledby="nationalities-select-label nationalities-select-checkbox"
        className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
        id="nationalities-select-checkbox"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        CH, ES, FR, GB
      </div>
      <input
        aria-hidden={true}
        className="MuiSelect-nativeInput"
        onAnimationStart={[Function]}
        onChange={[Function]}
        required={false}
        tabIndex={-1}
        value="CH,ES,FR,GB"
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
</div>
`;
